Data Cleaning

In [68]:
import pandas as pd
from sklearn.linear_model import Lasso
from sklearn.preprocessing import StandardScaler

In [69]:
df = pd.read_csv("NBA_combined_dataset.csv", encoding="latin1")
df.head()

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31,Guaranteed
0,A.J. Green,29.100000,26,MIL,SG,78,68,3.4,7.9,0.424,...,10.400000,NaN,MIL,2301587.0,10044644.0,10848215.0,11651786.0,12455355.0,NaN,47301587.0
1,Aaron Gordon,27.900000,30,DEN,PF,36,33,5.5,11.1,0.497,...,16.200001,NaN,DEN,22841455.0,33658037.0,36350680.0,39043323.0,NaN,NaN,92850172.0
2,Aaron Holiday,13.700000,29,HOU,PG,57,1,1.8,4.4,0.417,...,5.500000,NaN,HOU,2296274.0,NaN,NaN,NaN,NaN,NaN,2296274.0
3,Aaron Nesmith,29.700001,26,IND,SF,45,42,4.7,11.4,0.414,...,13.800000,NaN,IND,11000000.0,11000000.0,19418000.0,20971440.0,NaN,NaN,62389440.0
4,Aaron Wiggins,21.799999,27,OKC,SG,65,21,3.6,8.2,0.431,...,9.400000,NaN,OKC,10102803.0,9224300.0,8345797.0,8345797.0,NaN,NaN,27672900.0


In [70]:
df["Player"].value_counts()

Player
Garrison Mathews          3
Jeremiah Robinson-Earl    3
Kobe Bufkin               3
Tony Bradley              3
Jeremy Sochan             2
                         ..
Andrew Wiggins            1
Anfernee Simons           1
Anthony Black             1
Anthony Davis             1
Trayce Jackson-Davis      1
Name: count, Length: 405, dtype: int64

After checking for duplicate names, I realized that some players were listed in the dataset 2-3 times.

In [71]:
df[df["Player"] == "Garrison Mathews"]

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31,Guaranteed
131,Garrison Mathews,13.1,29,IND,SG,15,1,1.4,3.5,0.404,...,5.2,NaN,IND,429325.0,NaN,NaN,NaN,NaN,NaN,561295.0
132,Garrison Mathews,13.1,29,IND,SG,15,1,1.4,3.5,0.404,...,5.2,NaN,IND,429325.0,NaN,NaN,NaN,NaN,NaN,561295.0
133,Garrison Mathews,13.1,29,IND,SG,15,1,1.4,3.5,0.404,...,5.2,NaN,IND,429325.0,NaN,NaN,NaN,NaN,NaN,561295.0


In [72]:
df[df["Player"] == "Tony Bradley"]

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31,Guaranteed
388,Tony Bradley,11.0,28,2TM,C,41,4,1.6,2.9,0.547,...,4.0,NaN,IND,3204816.0,NaN,NaN,NaN,NaN,NaN,3204816.0
389,Tony Bradley,11.0,28,2TM,C,41,4,1.6,2.9,0.547,...,4.0,NaN,IND,3204816.0,NaN,NaN,NaN,NaN,NaN,3204816.0
390,Tony Bradley,11.0,28,2TM,C,41,4,1.6,2.9,0.547,...,4.0,NaN,IND,3204816.0,NaN,NaN,NaN,NaN,NaN,3204816.0


After further inspection for two players who are listed three times in the data, I found that all the rows for these specific players had the exact same data. Therefore, I decided to go ahead and drop duplicates to see if there are any players who have more than one row but different values in one or more columns.

In [73]:
df_cleaned = df.drop_duplicates()

In [74]:
df_cleaned["Player"].value_counts()

Player
Blake Wesley            2
Cole Anthony            2
Jeremy Sochan           2
Bradley Beal            2
Jevon Carter            2
                       ..
Andrew Wiggins          1
Anfernee Simons         1
Anthony Black           1
Anthony Davis           1
Trayce Jackson-Davis    1
Name: count, Length: 405, dtype: int64

Even after dropping duplicate rows, there were still some players that are represented more than once in the dataset.

In [75]:
df_cleaned[df_cleaned["Player"] == "Blake Wesley"]

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31,Guaranteed
32,Blake Wesley,11.6,22,POR,SG,31,0,1.8,4.0,0.452,...,4.8,NaN,WAS,5643732.0,NaN,NaN,NaN,NaN,NaN,3347458.0
33,Blake Wesley,11.6,22,POR,SG,31,0,1.8,4.0,0.452,...,4.8,NaN,POR,5643732.0,NaN,NaN,NaN,NaN,NaN,2296274.0


In [76]:
df_cleaned[df_cleaned["Player"] == "Jevon Carter"]

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31,Guaranteed
206,Jevon Carter,16.4,30,2TM,PG,53,1,2.3,5.8,0.401,...,6.4,NaN,CHI,7680524.0,NaN,NaN,NaN,NaN,NaN,6809524.0
207,Jevon Carter,16.4,30,2TM,PG,53,1,2.3,5.8,0.401,...,6.4,NaN,ORL,7680524.0,NaN,NaN,NaN,NaN,NaN,871000.0


Looking back at the NBA salary dataset, I noticed that some players had two entries with the only difference being different guaranteed contracts due to complex contracts with conditional guarantees. My goal for this project does not involve looking into future Guaranteed Contracts but instead current salary for the 2025-2026 season so I have decided to drop this column to easily delete duplicate players.

In [77]:
df_cleaned.drop("Guaranteed", axis=1, inplace=True)

C:\Users\mukun\AppData\Local\Temp\ipykernel_38156\294392823.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned.drop("Guaranteed", axis=1, inplace=True)


In [78]:
df_cleaned = df_cleaned.sort_values("Player").drop_duplicates("Player", keep="first")

In [79]:
df_cleaned["Player"].value_counts()

Player
A.J. Green         1
AJ Johnson         1
Aaron Gordon       1
Aaron Holiday      1
Aaron Nesmith      1
                  ..
Zach Edey          1
Zach LaVine        1
Zeke Nnaji         1
Ziaire Williams    1
Zion Williamson    1
Name: count, Length: 405, dtype: int64

In [80]:
df_cleaned[df_cleaned["Player"] == "Blake Wesley"]

,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PF_adv,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31
32,Blake Wesley,11.6,22,POR,SG,31,0,1.8,4.0,0.452,...,1.1,4.8,NaN,WAS,5643732.0,NaN,NaN,NaN,NaN,NaN


I have now cleaned the dataset of duplicate players. Now, I will look for missing values that may be concerning.

In [81]:
na_values = df_cleaned.isna().sum()
na_values[na_values > 1]

ThreePpct_pg      14
Awards_pg        354
ThreePpct_adv     14
Awards_adv       354
_2026_27         103
_2027_28         214
_2028_29         311
_2029_30         378
_2030_31         398
dtype: int64

For missing values, I noticed that a lot of the year columns had missing values, but this is expected as the player may not have a contract for upcoming year. Additionally, the awards section has missing values, but this also expected because most players do not recieve NBA awards. The only concerning column with missing values is 3pt percentage as this could potentially help in predicting NBA player salary.

In [82]:
df_cleaned[df_cleaned['ThreePpct_pg'].isna()]


,Player,MP,Age_pg,Team_pg,Pos_pg,G_pg,GS_pg,FG_pg,FGA_pg,FGpct_pg,...,PF_adv,PTS_adv,Awards_adv,Team_salary,_2025_26,_2026_27,_2027_28,_2028_29,_2029_30,_2030_31
39,Brandon Clarke,10.000000,29,MEM,PF,2,2,1.5,4.5,0.333,...,2.5,4.0,NaN,MEM,12500000.0,12500000.0,NaN,NaN,NaN,NaN
81,Daniel Gafford,21.700001,27,DAL,C,55,44,3.8,5.7,0.655,...,2.6,9.5,NaN,DAL,14386320.0,17263584.0,18126763.0,18989942.0,NaN,NaN
94,DeAndre Jordan,16.600000,37,NOP,C,12,7,1.8,2.7,0.656,...,1.3,4.4,NaN,NOP,2269880.0,NaN,NaN,NaN,NaN,NaN
91,Deandre Ayton,27.200001,27,LAL,C,72,72,5.6,8.3,0.671,...,2.2,12.5,NaN,POR,33654814.0,8104000.0,NaN,NaN,NaN,NaN
100,Dereck Lively II,16.400000,21,DAL,C,7,4,1.6,2.6,0.611,...,3.0,4.3,NaN,DAL,5253360.0,7239131.0,NaN,NaN,NaN,NaN
156,Ivica Zubac,30.100000,28,2TM,C,48,47,6.0,10.0,0.598,...,2.3,14.1,NaN,IND,18893980.0,20342140.0,21790300.0,NaN,NaN,NaN
165,Jakob Poeltl,25.000000,30,TOR,C,46,44,4.7,6.7,0.700,...,3.2,10.7,NaN,TOR,19500000.0,19500000.0,27300000.0,29484000.0,27300000.0,NaN
168,Jalen Duren,28.200001,22,DET,C,70,70,7.5,11.5,0.650,...,2.8,19.5,"DPOY-11,AS",DET,6483144.0,NaN,NaN,NaN,NaN,NaN
179,James Wiseman,14.500000,24,IND,C,4,1,1.5,2.5,0.600,...,2.5,3.3,NaN,IND,1131970.0,NaN,NaN,NaN,NaN,NaN
204,Jericho Sims,19.700001,27,MIL,C,67,19,2.2,2.8,0.784,...,2.2,5.0,NaN,MIL,2461463.0,2801346.0,NaN,NaN,NaN,NaN


In [83]:
df_cleaned["ThreePpct_pg"] = df_cleaned["ThreePpct_pg"].fillna(0)
df_cleaned["ThreePpct_adv"] = df_cleaned["ThreePpct_adv"].fillna(0)

In [84]:
br = df_cleaned[df_cleaned["Player"] == "Brandon Clarke"]
br["ThreePpct_pg"]

39    0.0
Name: ThreePpct_pg, dtype: float64

As I performed further research into players with missing 3pt percentages, I realized that these players have not shot a three pointer all season, which is why their 3pt percentage is missing. I decided to fill the null values in this column with 0 because if these players aren't shooting three pointers, they likely aren't good at 3pt shooting, making it fair to fill their 3pt percentage with 0 to represent that they are a bad shooter.

Lasso Regression

In order to fit an OLS model with high predictive probability, but low model complexity and multicollinearity I have decided to use LASSO Regression. Since there are 65 columns in the dataset, LASSO will help to cut out the least important and collinear predictor variables to ensure high model accuracy without too complex of a model.

Since LASSO models penalize having large slopes, I want to ensure that each of the slopes start with similar importance so that comparisons are fair. To do this, I will scale all the variables to their corresponding z-scores.

In [85]:
x_vars = df_cleaned.drop(columns= ["_2025_26", "Awards_pg","Awards_adv", "_2026_27", "_2027_28", "_2028_29", "_2029_30", "_2030_31" ])
x_numeric = x_vars.select_dtypes(include=['number'])
y_var = df_cleaned["_2025_26"]

In [86]:
x_numeric.isna().sum()

MP               0
Age_pg           0
G_pg             0
GS_pg            0
FG_pg            0
FGA_pg           0
FGpct_pg         0
ThreeP_pg        0
ThreePA_pg       0
ThreePpct_pg     0
TwoP_pg          0
TwoPA_pg         0
TwoPpct_pg       1
eFGpct_pg        0
FT_pg            0
FTA_pg           0
FTpct_pg         0
ORB_pg           0
DRB_pg           0
TRB_pg           0
AST_pg           0
STL_pg           0
BLK_pg           0
TOV_pg           0
PF_pg            0
PTS_pg           0
Age_adv          0
G_adv            0
GS_adv           0
FG_adv           0
FGA_adv          0
FGpct_adv        0
ThreeP_adv       0
ThreePA_adv      0
ThreePpct_adv    0
TwoP_adv         0
TwoPA_adv        0
TwoPpct_adv      1
eFGpct_adv       0
FT_adv           0
FTA_adv          0
FTpct_adv        0
ORB_adv          0
DRB_adv          0
TRB_adv          0
AST_adv          0
STL_adv          0
BLK_adv          0
TOV_adv          0
PF_adv           0
PTS_adv          0
dtype: int64

In [87]:
x_numeric = x_numeric.fillna(0)
x_numeric.isna().sum()

MP               0
Age_pg           0
G_pg             0
GS_pg            0
FG_pg            0
FGA_pg           0
FGpct_pg         0
ThreeP_pg        0
ThreePA_pg       0
ThreePpct_pg     0
TwoP_pg          0
TwoPA_pg         0
TwoPpct_pg       0
eFGpct_pg        0
FT_pg            0
FTA_pg           0
FTpct_pg         0
ORB_pg           0
DRB_pg           0
TRB_pg           0
AST_pg           0
STL_pg           0
BLK_pg           0
TOV_pg           0
PF_pg            0
PTS_pg           0
Age_adv          0
G_adv            0
GS_adv           0
FG_adv           0
FGA_adv          0
FGpct_adv        0
ThreeP_adv       0
ThreePA_adv      0
ThreePpct_adv    0
TwoP_adv         0
TwoPA_adv        0
TwoPpct_adv      0
eFGpct_adv       0
FT_adv           0
FTA_adv          0
FTpct_adv        0
ORB_adv          0
DRB_adv          0
TRB_adv          0
AST_adv          0
STL_adv          0
BLK_adv          0
TOV_adv          0
PF_adv           0
PTS_adv          0
dtype: int64

In [89]:
y_var.isna().sum()

np.int64(1)

In [90]:
y_var = y_var.fillna(0)

In [91]:
scaler = StandardScaler()
scaled_X = scaler.fit_transform(x_numeric)
X_scaled = pd.DataFrame(scaled_X, columns = x_numeric.columns)

Now, I can fit the LASSO Regression model.

In [102]:
lasso_mod = Lasso(alpha = 5000 , max_iter=5000)
lasso_mod.fit(X_scaled, y_var)
pd.DataFrame(lasso_mod.coef_, index = X_scaled.columns)


,0
MP,-2.880750e+06
Age_pg,3.996327e+06
G_pg,-8.976857e+05
GS_pg,1.311813e+06
FG_pg,-0.000000e+00
FGA_pg,-0.000000e+00
FGpct_pg,-0.000000e+00
ThreeP_pg,0.000000e+00
ThreePA_pg,-9.566585e+04
ThreePpct_pg,-5.623227e+05


In [103]:
lasso_mod.score(X_scaled,y_var)

0.6511526848059663

The LASSO Regression helped to limit the predictor variables I want to include in my linear model, while still producing a R-squared value of 0.6511526848059663

Multiple Linear Regression using Ordinary Least Squares